File: Problems/TwoPointersType/Is_SubSequence/is_subsequence.py
class Solution:
    # actually used two pointers thought implicitly:
    # one is i -> len(t)
    # another one is compare_idx
    def isSubsequence(self, s: str, t: str) -> bool:
        compare_idx: int = 0
        for i in range(len(t)):
            if compare_idx == len(s): return True
            if t[i] == s[compare_idx]:
                compare_idx += 1

        return True if compare_idx == len(s) else False   
   
if __name__ == "__main__":
    test: dict[tuple[str, str], bool] = {("abc","ahbgdc"): True, ("axc","ahbgdc"): False} 
    # use dict.items() extract all items in the dict
    for (key, val), flag in test.items():
        got: bool = Solution().isSubsequence(key, val)
        print(f"Got: {got} and expected: {flag}")

File: Problems/TwoPointersType/Is_SubSequence/IsSubSequence.java
package TwoPointersType.Is_SubSequence;


public class IsSubSequence {
    public static void main(String[] args) {

    }
}

class Solution {
    public boolean isSubsequence(String s, String t) {
        int compareIdx = 0;
        if (s.length() == 0) return true;

        for (int i = 0; i < t.length(); i++){
            if (compareIdx == s.length()) return true;
            if (s.charAt(compareIdx) == t.charAt(i)) compareIdx++;
        }
        return compareIdx == s.length();
    }
}
